Data/ByteString/Builder/Internal.hs

```diff
@@ -850,19 +850,17 @@ byteStringCopy = \bs -> builder $ byteStringCopyStep bs
 
 {-# INLINE byteStringCopyStep #-}
 byteStringCopyStep :: S.StrictByteString -> BuildStep a -> BuildStep a
-byteStringCopyStep (S.BS ifp isize) !k0 br0@(BufferRange op ope)
+byteStringCopyStep bs@(S.BS ifp isize) k br@(BufferRange op ope)
     -- Ensure that the common case is not recursive and therefore yields
     -- better code.
-    | op' <= ope = do copyBytes op ip isize
-                      touchForeignPtr ifp
-                      k0 (BufferRange op' ope)
-    | otherwise  = wrappedBytesCopyStep (BufferRange ip ipe) k br0
+       -- What's the reasoning here, more concretely?
```

Comment:
Are you going to elaborate on this?

Reply:
It's hard for me to elaborate. I thought about it for a while but still have _no idea_ in what sense a non-recursive hot path was supposed to be "better" at the time this was written.

* For small statically-known lengths, nowadays, we might benefit from specialized `copyBytes` code generation.
  * But `copyBytes` just lowered to an opaque foreign call when this comment was written.
  * This special code generation is not great until [ghc#24519](https://gitlab.haskell.org/ghc/ghc/-/issues/24519) is fixed.
  * Anyway, it'd be nice to somehow incur this extra branch only for literals if this is our reasoning.
* Recursive stuff is much less likely to inline automatically.
  * The fact that we need inlining of `b1` to happen to make a `Builder` concatenation `b1 <> b2` fast is pretty ugly.
    * I think on my machine the reboxing of args and the unknown call incur around 15ns of overhead per builder. There can also be a bit more overhead if the continuation has to allocate a PAP.
    * At the time the `Builder` core stuff was written, it was found that unboxing the `Builder` args actually made things slower, for obscure RTS fast-call-pattern reasons. (I think it was about equal when I tried it a year or two ago.)
  * If this is really that important, inlining of the whole loop could be forced with an inner-`go` and an `INLINE` pragma. And that entire loop isn't bigger than this one unrolled iteration!

I suppose in light of the lack of clear reasoning I'm fine with deleting this comment and unifying `byteStringCopyStep` with `wrappedBytesCopyStep`.

Reply:
I would also have to look at benchmarks, but we don't have any that really test this stuff right now. That changes in #569, but actually a change I wanted to make in that PR is blocked on this patch.

Reply:
It turns out I was wrong. There is a relevant benchmark already, `foldMap byteStringCopy`, and getting rid of the `byteStringCopyStep`/`wrappedBytesCopyStep` distinction slows that benchmark down by about 30% for some very obscure reasons. I'll push a comment describing these reasons in a few minutes.

This whole business with the continuation-threading causing practically unpredictable performance issues with `Builder`s has gotten very tiresome. Perhaps I will revisit #194 relatively soon.